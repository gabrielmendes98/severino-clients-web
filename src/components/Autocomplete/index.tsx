import { SyntheticEvent, useEffect, useMemo } from 'react';
import TextField, { TextFieldProps } from '@mui/material/TextField';
import { InputProps } from '@mui/material/Input';
import MuiAutocomplete, {
  AutocompleteProps as MuiAutocompleteProps,
} from '@mui/material/Autocomplete';
import debounce from 'lodash.debounce';

type Props = {
  id?: string;
  placeholder?: string;
  maxWidth?: number;
  onChange: (event: SyntheticEvent<Element, Event>, value: string) => any;
  options: SelectOptions;
  loading?: boolean;
  onOptionSelect?: (
    event: SyntheticEvent<Element, Event>,
    option: SelectOption,
  ) => any;
  InputProps?: InputProps;
  TextFieldProps?: TextFieldProps;
  name?: string;
  disableClearable?: boolean;
  freeSolo?: boolean;
};

interface AutocompleteProps<
  T,
  Multiple extends boolean | undefined,
  DisableClearable extends boolean | undefined,
  FreeSolo extends boolean | undefined,
> extends Omit<
      MuiAutocompleteProps<T, Multiple, DisableClearable, FreeSolo>,
      keyof Props | 'renderInput'
    >,
    Props {}

const Autocomplete = <
  T,
  Multiple extends boolean | undefined,
  DisableClearable extends boolean | undefined,
  FreeSolo extends boolean | undefined,
>({
  id,
  placeholder,
  maxWidth,
  onChange,
  onOptionSelect,
  options = [],
  InputProps,
  TextFieldProps,
  name,
  disableClearable = false,
  freeSolo = true,
  ...props
}: AutocompleteProps<T, Multiple, DisableClearable, FreeSolo>) => {
  const onType = useMemo(
    () => debounce((e, value) => onChange(e, value), 500),
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [],
  );

  const onOptionSelectHandler = (
    event: SyntheticEvent<Element, Event>,
    option: SelectOption | null | string,
  ) => {
    if (onOptionSelect && option) {
      (event.target as any).name = name;
      onOptionSelect(event, option as SelectOption);
    }
  };

  // eslint-disable-next-line react-hooks/exhaustive-deps
  useEffect(() => () => onType.cancel(), []);

  return (
    <MuiAutocomplete
      {...(props as any)}
      id={id || name}
      onInputChange={onType}
      freeSolo={freeSolo}
      options={options}
      onChange={onOptionSelectHandler}
      fullWidth
      sx={{
        maxWidth,
      }}
      disableClearable={disableClearable}
      renderInput={params => (
        <TextField
          {...params}
          size="small"
          placeholder={placeholder}
          InputProps={{
            ...params.InputProps,
            ...InputProps,
          }}
          name={name}
          {...TextFieldProps}
        />
      )}
    />
  );
};

export default Autocomplete;
