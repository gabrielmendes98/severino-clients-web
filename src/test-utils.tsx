import React, { FC, ReactElement, ReactNode } from 'react';
import { render, RenderOptions } from '@testing-library/react';
import { Provider } from 'react-redux';
import ThemeProvider from './common/providers/Theme';
import { makeStore } from './common/store/store';

// const Wrapper: FC<{ children: React.ReactNode }> = ({ children }) => (
//   <ThemeProvider>{children}</ThemeProvider>
// );

// const customRender = (
//   ui: ReactElement,
//   options?: Omit<RenderOptions, 'wrapper'>,
// ) => render(ui, { wrapper: wrapper.withRedux(Wrapper), ...options });

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, 'wrapper'>,
) => {
  const Wrapper: FC<{ children: ReactNode }> = ({ children }) => (
    <Provider store={makeStore({})}>
      <ThemeProvider>{children}</ThemeProvider>
    </Provider>
  );

  return { ...render(ui, { wrapper: Wrapper, ...options }) };
};

/* eslint-disable */
export * from '@testing-library/react';
export { default as userEvent } from '@testing-library/user-event';
export { customRender as render };
/* eslint-enable */
