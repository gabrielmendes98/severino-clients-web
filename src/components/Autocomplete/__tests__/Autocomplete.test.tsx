import { render, userEvent, screen, waitFor, within } from 'test-utils';
import Autocomplete from '../index';

const mockOptions = [
  { label: 'option one', value: '1' },
  { label: 'option two', value: '2' },
];

it('should match snapshot', () => {
  const { container } = render(
    <Autocomplete onChange={jest.fn()} options={mockOptions} />,
  );
  expect(container).toMatchSnapshot();
});

it('should call onChange with event and value', async () => {
  const onChange = jest.fn();
  render(
    <Autocomplete
      onChange={onChange}
      options={mockOptions}
      name="autocomplete"
      TextFieldProps={{ label: 'autocomplete input' }}
    />,
  );
  userEvent.type(screen.getByLabelText('autocomplete input'), 'foo');
  await waitFor(() => {
    expect(onChange).toHaveBeenCalledWith(expect.any(Object), 'foo');
  });
});

it('should call onOptionSelect when select option', async () => {
  const onChange = jest.fn();
  const onOptionSelect = jest.fn();
  render(
    <Autocomplete
      onChange={onChange}
      options={mockOptions}
      name="autocomplete"
      TextFieldProps={{ label: 'autocomplete input' }}
      onOptionSelect={onOptionSelect}
    />,
  );
  userEvent.type(screen.getByLabelText('autocomplete input'), 'option');
  const list = await screen.findByRole('listbox');
  const optionOne = within(list).getByText('option one');
  userEvent.click(optionOne);
  await waitFor(() => {
    expect(onOptionSelect).toHaveBeenCalledWith(
      expect.any(Object),
      mockOptions[0],
    );
  });
});
