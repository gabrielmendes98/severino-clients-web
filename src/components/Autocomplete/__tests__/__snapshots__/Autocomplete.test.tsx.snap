// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<div>
  <div
    class="Toastify"
  />
  <div
    class="MuiAutocomplete-root MuiAutocomplete-fullWidth css-eozd4h-MuiAutocomplete-root"
  >
    <div
      class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
    >
      <div
        class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-sizeSmall MuiAutocomplete-inputRoot css-3xjrk5-MuiInputBase-root-MuiOutlinedInput-root"
      >
        <input
          aria-autocomplete="list"
          aria-expanded="false"
          aria-invalid="false"
          autocapitalize="none"
          autocomplete="off"
          class="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputSizeSmall MuiAutocomplete-input MuiAutocomplete-inputFocused css-1n4twyu-MuiInputBase-input-MuiOutlinedInput-input"
          id=":r0:"
          role="combobox"
          spellcheck="false"
          type="text"
          value=""
        />
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-hdw1oc"
          >
            <span
              class="notranslate"
            >
              â€‹
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
  </div>
</div>
`;
